File: src/game/ts/runner.ts
import {
	CellType,
	ChainReaction,
	ChainReactionOptions,
} from "./chain_reaction";
import { array, CancelPromise } from "@common/util";
import { Hooks, extendHooks } from "@common/hooks";
import { neighbors, XY } from "./common";

export interface GameContext {
	width: ChainReaction["width"];
	height: ChainReaction["height"];
	grid: ChainReaction["grid"];

	player: number;
	emptyCells: Set<number>;
	ownedCells: Set<number>[];
	availableCells(): Set<number>;

	canPlace: ChainReaction["canPlace"];
	neighbors(pos: number): number[];
	mass(pos: number): number;
	capacity(pos: number): number;
}

export interface Playable {
	play(context: GameContext): XY | Promise<XY>;
}

interface RunnerOptions extends Omit<ChainReactionOptions, "players"> {
	players: Playable[];
}

function createGameContext(player: number, game: ChainReaction): GameContext {
	const { width, height, emptyCells, ownedCells, grid } = game;

	return Object.freeze({
		width,
		height,
		grid,
		player,
		emptyCells,
		ownedCells,
		availableCells() {
			return new Set([...game.emptyCells, ...game.ownedCells[player]]);
		},

		canPlace: game.canPlace.bind(game),
		capacity: game.capacity.bind(game),

		neighbors(pos) {
			return neighbors(pos, width, height);
		},

		mass(pos) {
			const cell = grid[pos];
			return cell.type === CellType.Owned ? cell.count : 0;
		},
	});
}

export class Runner {
	readonly game: ChainReaction;
	private players: Playable[];

	/** Gets passed to players so they can decide what their move will be. */
	private gameContext: Readonly<GameContext>[];

	private cancelPromise = CancelPromise();
	private tallyPromise = CancelPromise<number[]>();
	hooks: Hooks<"update" | "turnDelay" | "explosionDelay" | "gameDelay">;

	constructor(options: RunnerOptions) {
		const { width, height, players } = options;

		this.players = players;
		this.game = new ChainReaction({ width, height, players: players.length });
		this.gameContext = array(players.length, player =>
			createGameContext(player, this.game),
		);

		this.hooks = extendHooks(this.game.hooks, "turnDelay", "gameDelay");
	}

	/**
	 * Plays out the game until completion the given number of times.
	 *
	 * @returns An array representing a tally of the players' victories.
	 */
	private async _run(
		times: number,
		onGameFinished: (winner: number, id: number, tally: number[]) => void,
	): Promise<number[]> {
		const tally = array(this.players.length, () => 0);

		for (let i = 1; i <= times; i++) {
			while (this.game.isActive()) {
				if (this.cancelPromise.cancelled) {
					await this.game.reset();
					return tally;
				}

				const player = this.players[this.game.currentPlayer];

				const gameContext = this.gameContext[this.game.currentPlayer];

				const availableCells = new Set([
					...gameContext.ownedCells[this.game.currentPlayer],
					...gameContext.emptyCells,
				]);

				if (
					this.game.playerIsAlive(this.game.currentPlayer) &&
					availableCells.size
				) {
					const pos = await Promise.race([
						player.play(gameContext),
						this.cancelPromise.promise as Promise<void>,
					]);

					if (pos) {
						await this.game.place(pos.x, pos.y);
						await this.hooks.run("turnDelay", this.cancelPromise.promise);
					}
				} else {
					// Skip the player's turn if they can't make any legal moves.
					this.game.nextPlayer();
				}
			}

			if (!this.cancelPromise.cancelled) {
				const winner = this.game.winner();
				tally[winner] += 1;
				onGameFinished(winner, i, tally);
			}

			await this.hooks.run("gameDelay", this.cancelPromise.promise);
			await this.game.reset();
		}

		return tally;
	}

	run(
		times: number,
		onGameFinished: (winner: number, id: number, tally: number[]) => void,
	): CancelPromise<number[]> {
		this.cancelPromise = CancelPromise();
		this.tallyPromise = CancelPromise(this._run(times, onGameFinished));

		const oldCancel = this.tallyPromise.cancel.bind(this.tallyPromise);
		this.tallyPromise.cancel = () => {
			oldCancel();
			this.cancelPromise.cancel();
		};

		void this.tallyPromise.promise.then(() => {
			this.game.cancel();
			this.hooks.clear();
		});

		return this.tallyPromise;
	}
}

File: src/game/rust/src/runner.rs
use crate::array2d::{Array2D, Pos};
use crate::chain_reaction::{ChainReaction, Field};
use std::{
    cell::{Cell, RefCell},
    collections::HashSet,
    convert::TryInto,
};

pub trait Player {
    fn play(&mut self, ctx: GameContext) -> Result<Pos, &'static str>;
}

pub struct GameContext<'a> {
    game: &'a ChainReaction,
}

impl<'a> GameContext<'a> {
    pub fn new(game: &'a ChainReaction) -> Self {
        GameContext { game }
    }

    pub fn width(&self) -> usize {
        self.game.width()
    }

    pub fn height(&self) -> usize {
        self.game.height()
    }

    pub fn player(&self) -> usize {
        self.game.current_player()
    }

    pub fn grid(&self) -> &Array2D<Cell<Field>> {
        &self.game.grid
    }

    pub fn empty_cells(&self) -> &RefCell<HashSet<Pos>> {
        &self.game.empty_cells
    }

    pub fn owned_cells(&self) -> &RefCell<Vec<HashSet<Pos>>> {
        &self.game.owned_cells
    }

    pub fn available_cells(&self) -> HashSet<Pos> {
        let player = self.game.current_player();

        let empty = self.game.empty_cells.borrow();
        let available = self.game.owned_cells.borrow();

        empty
            .union(available.get(player).unwrap())
            .map(|x| *x)
            .collect()
    }

    pub fn capacity(&self, pos: &Pos) -> Result<&u32, &'static str> {
        self.game.capacity.get(pos)
    }

    pub fn mass(&self, pos: &Pos) -> Result<u32, &'static str> {
        let field = self.game.grid.get(pos)?;
        let field = field.get();

        let mass = match field {
            Field::Empty => 0,
            Field::Owned(data) => data.count,
        };

        Ok(mass)
    }

    pub fn neighbors(&self, pos: &Pos) -> Vec<(Pos, &Cell<Field>)> {
        self.game.neighbors(pos)
    }
}

pub struct Runner<'a, P>
where
    P: Player,
{
    width: usize,
    height: usize,
    game: ChainReaction,
    players: Vec<P>,

    on_game_finished: Option<Box<dyn Fn(&Vec<usize>, &usize, &u32) + 'a>>,
    should_stop: Option<Box<dyn Fn() -> bool + 'a>>,
}

impl<'a, P> Runner<'a, P>
where
    P: Player,
{
    pub fn new(
        width: usize,
        height: usize,
        players: Vec<P>,
        on_game_finished: Option<Box<dyn Fn(&Vec<usize>, &usize, &u32) + 'a>>,
        should_stop: Option<Box<dyn Fn() -> bool + 'a>>,
    ) -> Result<Runner<'a, P>, &'static str> {
        let game = ChainReaction::new(width, height, players.len())?;

        Ok(Runner {
            width,
            height,
            game,
            players,

            on_game_finished,
            should_stop,
        })
    }

    pub fn run_js(
        &mut self,
        times: u32,
        id_array: &js_sys::Uint32Array,
        tally_array: &js_sys::Uint32Array,
    ) -> Result<Vec<usize>, &'static str> {
        let mut tally = vec![0; self.players.len()];
        for id in 1..times + 1 {
            while self.game.active() {
                let player = self.game.current_player();
                let context = GameContext::new(&self.game);
                let player_move = self.players[player].play(context)?;
                self.game.place(&player_move)?;
            }

            let winner = self.game.winner()?;
            *tally.get_mut(winner).unwrap() += 1;
            let winner_u32: u32 = winner.try_into().unwrap();
            tally_array.set_index(winner_u32, tally_array.get_index(winner_u32) + 1);
            id_array.set_index(0, id);

            if id_array.get_index(1) != 0 {
                return Ok(tally);
            }

            self.game = ChainReaction::new(self.width, self.height, self.players.len())?;
        }
        Ok(tally)
    }

    pub fn run(&mut self, times: u32) -> Result<Vec<usize>, &'static str> {
        let mut tally = vec![0; self.players.len()];
        for id in 1..times + 1 {
            while self.game.active() {
                let player = self.game.current_player();
                let context = GameContext::new(&self.game);
                let player_move = self.players[player].play(context)?;
                self.game.place(&player_move)?;
            }

            let winner = self.game.winner()?;
            *tally.get_mut(winner).unwrap() += 1;

            if let Some(on_game_finished) = &self.on_game_finished {
                on_game_finished(&tally, &winner, &id);
            }

            if let Some(should_stop) = &self.should_stop {
                if should_stop() {
                    return Ok(tally);
                }
            }

            self.game = ChainReaction::new(self.width, self.height, self.players.len())?;
        }
        Ok(tally)
    }
}
